routers/web/auth/oauth.go

```diff
@@ -952,21 +952,30 @@ func SignInOAuthCallback(ctx *context.Context) {
 			if gothUser.Email == "" {
 				missingFields = append(missingFields, "email")
 			}
-			if setting.OAuth2Client.Username == setting.OAuth2UsernameNickname && gothUser.NickName == "" {
-				missingFields = append(missingFields, "nickname")
+			uname, err := extractUserNameFromOAuth2(&gothUser)
+			if err != nil {
+				ctx.ServerError("UserSignIn", err)
+				return
+			}
+			if uname == "" {
+				if setting.OAuth2Client.Username == setting.OAuth2UsernameNickname {
+					missingFields = append(missingFields, "nickname")
+				} else if setting.OAuth2Client.Username == setting.OAuth2UsernamePreferredUsername {
+					missingFields = append(missingFields, "preferred_username")
+				} // else: "UserID" and "Email" have been handled above separately
```

Comment:
I can not understand the old logic. It seems that `sub` and `email` are required fields and can not be empty.
If `setting.OAuth2Client.Username` is `nickname`, and oauth2 provider don't provide `email` field or the user's email is not set. (maybe this is not allowed by the rule of oauth2? not sure)
`len(missingFields)` will always greater than 0, then we will always print an error log below?

Reply:
> It seems that `sub` and `email` are required fields and can not be empty.

For "auto registration = true": yes, there are required.


> If `setting.OAuth2Client.Username` is `nickname`, and oauth2 provider don't provide `email` field or the user's email is not set. (maybe this is not allowed by the rule of oauth2? not sure)

`sub` (username) is a basic field of oauth2. `email` is required for "auto registration = true". `nickname` or other fields could be used as Gitea's username

> `len(missingFields)` will always greater than 0, then we will always print an error log below?

If any required field is missing:

* Before: 500 error, user panics.
* This PR: redirect to "link account" page and show a friendly message, and yes, by the way print an error log.

Reply:
Thanks for your answer.
Maybe we can adjust the error message? Maybe these fields are provided but values are empty?

Reply:
No, they shouldn't be empty. If these fields are required, they can't be empty, the error message is right.